overlord/snapshotstate/backend/backup.go

```diff
@@ -0,0 +1,62 @@
+package backend
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+// Backup stores information that can be used to cleanly revert (or
+// finish cleaning up) a snapshot Restore.
+//
+// This is useful when a Restore is part of a chain of operations, and
+// a later one failing necessitates undoing the Restore.
+type Backup struct {
+	Done    bool     `json:"done,omitempty"`
+	Created []string `json:"created,omitempty"`
+	Moved   []string `json:"moved,omitempty"`
+	// Config is here for convenience; this package doesn't touch it
+	Config *json.RawMessage `json:"cfg,omitempty"`
+}
+
+// check that you're not trying to use backup info twice; data loss imminent.
+func (b *Backup) check() {
+	if b.Done {
+		panic("attempting to use a snapshot.Backup twice")
+	}
+	b.Done = true
+}
+
+// Cleanup the backed up data from disk.
+func (b *Backup) Cleanup() {
+	b.check()
+	for _, dir := range b.Moved {
+		if err := os.RemoveAll(dir); err != nil {
+			logger.Noticef("cannot remove directory tree rooted at %q: %v", dir, err)
+		}
+	}
+}
+
+// Revert the backed up data: remove what was added, move back what was moved aside.
+func (b *Backup) Revert() {
+	b.check()
+	for _, dir := range b.Created {
+		logger.Debugf("removing %q", dir)
```

Comment:
Not super important but in a sea of generic messages we could say `removing backup %q`

Reply:
it'll already say `backup.go`, are you sure you need the extra bit?